393 cbyrd vue component guidance (#394)
archesproject/arches-docs: docs/developing/vue/arches-vue-integration.md

import MyVueApplication from '@/MyVueApplication.vue';

createVueApplication(MyVueApplication).then(vueApp => {
    vueApp.mount('#my-vue-application-mounting-point');
});
```

In the provided example, the createVueApplication function, imported from `utils/create-vue-application`, streamlines the integration process by handling the necessary setup and configuration steps, allowing developers to focus on building the core functionality of their Vue application.

The `createVueApplication` function takes a Vue component (`MyVueApplication` in this case) as its argument, representing the root component of the Vue application. This component encapsulates the entire application logic and user interface.

Once the Vue application is created using `createVueApplication`, it returns a Vue application instance (`vueApp`), which can then be further manipulated or customized as needed. In the example, the `vueApp` instance is mounted to a specific element in the DOM with the mount method, using the CSS selector `#my-vue-application-mounting-point` to identify the mounting point.

## Single-Responsibility Principle and Component Decomposition

> In Vue development, it's crucial to adhere to the Single Responsibility Principle (SRP) and practice component decomposition to ensure that Vue components remain maintainable and scalable. The SRP dictates that each Vue component should have a single responsibility or purpose. By focusing on doing one thing and doing it well, components become easier to understand, modify, and maintain. Following the SRP leads to more modular, reusable, and testable code. 

Reviewer: (side note: I tend to decompose more aggressively toward the end of a feature once all of my original assumptions have had the opportunity to be violated three or four times 😉 , but I agree with the sentiment I'm detecting here which is that this is perhaps *the* most important thing in this style guide!)

Component decomposition involves breaking down complex Vue components into smaller, more focused units, with each unit responsible for a specific task or feature. This practice aligns with the SRP and promotes clean, maintainable codebases.

Example:

```
<script setup>
import { ref } from 'vue';
import UserProfileHeader from './UserProfileHeader.vue';
import UserProfileBio from './UserProfileBio.vue';

> import UserProfileActions from './UserProfileActions.vue';

Reviewer: Do you like to alphabetize imports? It's totally fair if the answer is "no, not unless a pre-commit tool does it for me 😄 ", but I wonder if we might just alphabetize this one in the doc (within this group of 3 first-party imports).

Author: Agreed! I prefer to alphabetize imports && group them by source, and agreed that they should be alphabetized for the doc.

If there's a pre-commit hook that can easily be configured to handle this, the better. Otherwise there's probably too much friction for proper encforcement.

const user = ref({
  name: 'John Doe',
  email: 'john@example.com',

>   bio: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.'

Reviewer: ```suggestion
  bio: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
```

});
</script>

<template>

>   <div>

Reviewer: I'd love to get a `.prettierrc` in place as soon as possible so we can standardize on:
- single vs. double quotes
- 2 vs. 4 spaces
- line-length

And have a github action step fail PRs that haven't been formatted. Then we can socialize the knowledge about how to configure "format on save" in VS Code so that you never have to hand-format again. (Separately, I'll suggest doing the same thing with the python backend because we seem to get lots of PRs with unrelated formatting changes on a regular basis...)

I don't have a strong preference about quote or indentation width myself.

Reviewer: archesproject/arches#10575

- Each smaller component has a specific responsibility:
    - `UserProfileHeader`: Renders the user's name and email.
    - `UserProfileBio`: Renders the user's bio.
    - `UserProfileActions`: Renders user actions (e.g., buttons for editing the profile).
- The user reactive reference is defined using ref() within the `<script setup>` block and is accessible to all components within the template.
- The `UserProfile` component integrates these smaller components directly in the template.
- By decomposing the `UserProfile` component into smaller, focused components, we achieve better maintainability and reusability in our Vue application.

## Directory Structure

A well-organized directory structure provides clarity on where to find specific files and components, making it easier for developers to navigate the codebase, understand its architecture, and make modifications efficiently. In Arches Vue applications, we utilize a non-standard directory structure to organize components. 

Example:

- src/
    - components/
        - UserProfile/
            - UserProfileHeader.vue
            - UserProfileBio.vue
            - UserProfileActions.vue
            - UserProfile.vue

>         - Map/

Reviewer: Should we include an example with a nested `/types` directory? When typing an API response (in particular) it's nice to have them outside somewhere that don't clutter up the component. Or is that overkill for this example.

            - MapHeader.vue
            - MapSidebar.vue
            - Map.vue
    - views

>         - UserProfileView.vue

Reviewer 2: If Header, Bio, and Actions are all imported into UserProfile.vue what goes in UserProfileView.vue?

Author: :wave: @ryan86 

`UserProfileView.vue` is a higher-order component. It's purpose is to bring together the lower-order components and maybe some window dressing around them. eg

```
<script setup>
import { ref } from 'vue';
import UserProfileHeader from './UserProfileHeader.vue';
import UserProfileBio from './UserProfileBio.vue';
import UserProfileActions from './UserProfileActions.vue';

const user = ref({
  name: 'John Doe',
  email: 'john@example.com',
  bio: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.'
});
</script>

<template>
  <div>
    <UserProfileHeader :name="user.name" :email="user.email" />
    <UserProfileBio :bio="user.bio" />
    <UserProfileActions />
  </div>
</template>
```

    - reports
        - MapReport.vue
    - widgets
        - MapWidget.vue

> As Arches moves forward with integrating components from Vue into Knockout, we find it essential to adopt a non-standard directory structure to accommodate this scenario. For instance it's standard practice for most Vue applications to have a top-level `App.vue` file which has one or many child components that exist in the `components` directory. However since Arches will have multiple Vue applications running per project, possibly even per page, there must be some deliniation between these top-level components. The pattern exemplified above illustrates a way to have such deliniation while maintaining a scalable and navigatiable structure. Essentially, top-level components live in `views`, `reports`, `widgets`, etc, while their child components live in the `components` directory. The `components` subdirectories each contain a root component which is named the same as the subdirectory.

Reviewer: ```suggestion
As Arches moves forward with integrating components from Vue into Knockout, we find it essential to adopt a non-standard directory structure to accommodate this scenario. For instance it's standard practice for most Vue applications to have a top-level `App.vue` file which has one or many child components that exist in the `components` directory. However since Arches will have multiple Vue applications running per project, possibly even per page, there must be some delineation between these top-level components. The pattern exemplified above illustrates a way to have such delineation while maintaining a scalable and navigable structure. Essentially, top-level components live in `views`, `reports`, `widgets`, etc, while their child components live in the `components` directory. The `components` subdirectories each contain a root component which is named the same as the subdirectory.
```

> As Arches moves forward with integrating components from Vue into Knockout, we find it essential to adopt a non-standard directory structure to accommodate this scenario. For instance it's standard practice for most Vue applications to have a top-level `App.vue` file which has one or many child components that exist in the `components` directory. However since Arches will have multiple Vue applications running per project, possibly even per page, there must be some deliniation between these top-level components. The pattern exemplified above illustrates a way to have such deliniation while maintaining a scalable and navigatiable structure. Essentially, top-level components live in `views`, `reports`, `widgets`, etc, while their child components live in the `components` directory. The `components` subdirectories each contain a root component which is named the same as the subdirectory.

Reviewer: This example works really well. I wonder if we even need to bother calling it nonstandard. I instinctively put my Controlled List Manager components under `/components/ControlledListManager/`, and left a general-purpose spinner outside under `/components/Spinner.vue`.

Given that, what do you think about general-purpose UI elements like `<Spinner>`. Should we have a folder for things like that?

Author: In my mind, `Spinner` is not a top-level component and therefore belongs squarely in the `components` directory. However definitely happy to change the idea if the larger group arrives at a consensus. A larger topic this touches on is getting rid of the page-level spinner for component-level shimmers/spinners. Regardless they could all be powered by the same generic component.

### Enforcement

While this integration brings significant benefits to our Vue projects, it's important to note that newly written components will be affected by TypeScript and ESLint linting rules. Developers will need to adhere to TypeScript typing conventions and ESLint rules when writing new components to ensure consistency and compliance with the established coding standards. This enforcement takes place in the `build_development`, `build_test`, and `build_production` yarn processes. Previously written components will not be affected.

We have also added the `eslint:check`, `eslint:watch`, `ts:check`, and `ts:watch` yarn scripts.

### Separate Processes

During development, developers should run the TypeScript and ESLint `:watch` linters in separate processes to ensure efficient linting of the current project. By running these linters in separately, each process is dedicated to linting the current project or application. However, it's important to note that if you're running multiple projects or applications concurrently, you'll need a separate process for each project to ensure accurate linting results. 

> For instance, if you're developing an Arches project and an Arches application, you will now likely have six concurrent processes: The project-level Django server, the project-level webpack development server, the project-level typescript watcher, the project-level eslint watcher, the application-level typescript watcher, and the application-level eslint watcher.

Reviewer: I'm glad this is called out, because I can foresee forum posts along the lines of "why didn't I see ts warnings"? However, in the event this sounds a little overwhelming to some folks, what do you think about mentioning configuring your IDE as an alternative to running a watcher.

Vite [docs](https://vitejs.dev/guide/features.html#transpile-only) on "assum[ing] type checking is taken care of by your IDE and build process."

Author: For sure agreed on the alternative mention. Also yeah should add a mention about how the linter/TS checker have been built into the build processes so it's not necessary to run a separate process

### Frontend dependency declaration

> When adding any new frontend dependencies to your Arches project or application, it's important to declare their types in the `declarations.d.ts` file. This file serves as a central location for declaring global types and interfaces that are used across your project or application. Declaring types for frontend dependencies in `declarations.d.ts` ensures that TypeScript recognizes and understands the types provided by these modules, enabling accurate type checking and providing better IntelliSense support in your codebase. 

Reviewer: ```suggestion
When adding any new frontend dependencies to your Arches project or application, it's important to declare their types in the `declarations.d.ts` file. This file serves as a central location for declaring global types and interfaces that are used across your project or application. Declaring types for frontend dependencies in `declarations.d.ts` ensures that TypeScript recognizes and understands the types provided by these modules, enabling accurate type checking and providing better IntelliSense support in your editor. 
```

Example:
```
# package.json
    ...
    "dependencies": {
        "foo": "0.0.1",
        ...
    }
    ...

# src/declarations.d.ts
    ...
    declare module 'foo';

```

### Referencing `this` inside Vue `<template>` tags

> When accessing data properties, computed properties, or methods within a Vue component's `<template>` tags, it's essential to explicitly reference `this`. This ensures that TypeScript understands the context and correctly types the data or methods. Failing to use explicit `this` references inside `<template>` tags will result in TypeScript errors.

Reviewer: Noticed we'll have to silence the vue warnings for this:
```
[Vue warn]: Property 'items' was accessed via 'this'. Avoid using 'this' in templates. 
```

And it's too bad that JS variables in `<script setup>` now show as unused in VS Code. Maybe the tooling will catch up.

In vue 3.4+ (once the tooling catches up, b/c I don't seem to be able to use it now), we might be able to prefer:
```
:items
```
instead of:
```
:items="this.items"
```
Thoughts?

## Composition API and Single-file Components

We strongly suggest utilizing the Composition API and Single-file Components for building new features or rewriting existing components into Vue.

### Composition API

The Composition API is a way of organizing and reusing logic within Vue components. It allows developers to encapsulate related logic into reusable composition functions, making it easier to manage complex component logic and share code between components.

### Single-file Components

Single-file Components are a feature of Vue that allows developers to define templates, script, and styles in a single `.vue` file. This approach promotes a more modular and cohesive structure for Vue components, making it easier to manage component-specific logic, styles, and templates in a single file. 

### Example
Here's an example of a Single-file Component that uses the Composition API, notice the <script>, <template>, and <style> tags for a component exist in the same file:

```
<script setup lang="ts">
import { ref } from 'vue';

const count = ref(0);
const increment = () => {
  count.value++;

> };

Reviewer: This example is perfect. There's nothing else that makes sense.

In a more contrived example, you end up with more options about where to shove things. If there are a lot of arrow functions, you can group them all together, or you can keep the setter right next to where the state is declared. (Kind of like `const [count, setCount] = useState(0);` in React.)

In other words, a more contrived example might look like (knowing full well that the worse this gets the more it signals some decomposition is called for 😉 ):
```vue
<script setup lang="ts">
import { ref } from 'vue';  // third-party import
...

import { FancyButton } from 'primevue'; // will be a handful of these
...

import "MyComponent" from "@...";  // first-party import
...

import type { ApiResponse } from "@types/...'; // types
...

const LIMIT = 42; // constants

const [prop1, prop2]: {prop1: number, prop2: string} = defineProps(["prop1", "prop2"]);  // props
const toast = useToast(); // hooks
const count = ref(0);  // state
const otherReactiveState = ref([]);  // even more state
...

const formattedValue = computed(() => { // this is a lot like state...
...
});

const increment = () => { // begin arrow functions....
...

const onClick = () => {
...

const onReorder = () => {
...

const fetchData = async () => {
...

await fetchData();  // initialization code
```

Again, this is testing the edges of too-much-in-one-file, but I think it's not far from realistic (a couple of pieces of reactive state, a couple of callbacks, data fetching, and a toast for error handling)

These are all debatable, but what do you think about:
- Imports at top, whitespace between 3rd party and 1st party, optionally around primevue imports
- Import types last
- static constants, then props, then reactive state, then computed/watch, etc.
- functions mutating state declared right after state
- otherwise arrow functions toward the bottom
- any calls to those functions at the very end

Reviewer 2: Should each category also be alphabetized? Static consts A-Z, props A-Z, functions A-Z, etc?

Reviewer 2: Also, maybe additional category for vue lifecycle hooks unless we want to include them with arrow functions?

Reviewer: @ryan86 :wave:

alpha'ing constants and callbacks seems like overkill to me in my experience. 🤷‍♂️

</script>

<template>
  <button @click="this.increment">Increment</button>
  <p class="counter">Count: {{ this.count }}</p>
</template>

<style scoped>
.counter {
    color: red;
}
</style>

```

## Testing

TBD - This will be included once a testing pattern has been explored.

## Example Arches Vue Component Integration

> Below is an example of creating a new plugin for Arches. This example includes import patterns, TypeScript patters, internationalization, the composition API, single-file components, and Vue/Knockout integration

Reviewer: ```suggestion
Below is an example of creating a new plugin for Arches. This example includes import patterns, TypeScript patterns, internationalization, the composition API, single-file components, and Vue/Knockout integration.
```

```
// media/js/views/components/plugins/my-plugin.js

import ko from 'knockout';
import MyPlugin from '@/plugins/MyPlugin.vue';
import createVueApplication from 'utils/create-vue-application';
import MyPluginTemplate from 'templates/views/components/plugins/my-plugin.htm';

ko.components.register('my-plugin', {
    viewModel: function() {

>         createVueApplication(MyPluginVueApp).then(vueApp => {

Reviewer: ```suggestion
        createVueApplication(MyPlugin).then(vueApp => {
```

            vueApp.mount('#my-plugin-mounting-point');
        });
    },
    template: MyPluginTemplate,
});

```

```
// templates/views/components/plugins/my-plugin.htm

<div id="my-plugin-mounting-point"></div>
```

```
// src/plugins/MyPlugin.vue

<script setup lang="ts">
import { useGettext } from 'vue3-gettext';
const { $gettext } = useGettext();

console.log($gettext('Foo!'))
</script>

<template>

>     <h1 class="foo">{{ this.$gettext("Bar!") }}</h1>

Reviewer: Nice example. Foo!

[_Off topic_] By the way, do you happen to know what procedure compiles strings for translators? Do they get de-duped between the frontend and backend the way they do now? E.g.:
```
arches/location1.py
arches/location2.py
arches/app/media/js/blah.js
arches/app/src/components/BarReport/BarReportHeader.vue
arches_for_science/plugins/whatever.py
===
BAR!
```

Author: :wave: @jacobtylerwalls 

As far as the `vue3-gettext` implementation is concerned, it's on a per-filetree/project basis. Essentially the `path` value in `gettext.config.js` can only accept a single filepath. I've not dug in enough to know if there's de-duplication at the filetree/project level but that will be interesting to explore
